File: Python/0002_AddTwoNumbers/AddTwoNumbers.cpp
#pragma once
#include <cstddef>

struct ListNode {
	int val;
	ListNode* next;

	ListNode(int x) : val(x), next(NULL) {
	}
};

class Solution {
public:
	ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
		ListNode* head = new ListNode(0);
		ListNode* ans = head;
		int carry = 0;
		int sumVal;
		while (l2 != NULL && l1 != NULL) {
			sumVal = l1->val + l2->val + carry;
			carry = sumVal >= 10;			
			ans->next = new ListNode(carry ? sumVal - 10 : sumVal);
			l1 = l1->next;
			l2 = l2->next;
			ans = ans->next;
		}
		ListNode* l3 = l1 == NULL ? l2 : l1;
		while (l3 != NULL) {
			sumVal = l3->val + carry;
			carry = sumVal >= 10;
			ans->next = new ListNode(carry ? sumVal - 10 : sumVal);
			l3 = l3->next;
			ans = ans->next;
		}
		if (carry)
			ans->next = new ListNode(1);
		return head->next;

	}
};

File: Python/0002_AddTwoNumbers/AddTwoNumbers.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        # 72ms, 13.6MB
        [a, b] = self.oneBitSum(l1.val + l2.val)
        sumL = ListNode(a)
        head = ListNode(0)
        head.next = sumL
        l1, l2 = self.nextNode(l1, l2)
        while l1 is not None and l2 is not None:
            [a, b] = self.oneBitSum(l1.val+l2.val+b)
            sumL.next = ListNode(a)
            sumL, l1 = self.nextNode(sumL, l1)
            l2 = l2.next
        if l1 is None:
            sumL.next = l2
        else:
            sumL.next = l1
        if b == 0:
            return head.next
        while sumL.next is not None:
            sumL = sumL.next
            [a, b] = self.oneBitSum(sumL.val+b)
            sumL.val = a
            if b == 0:
                return head.next
        if b != 0:
            sumL.next = ListNode(1)
        return head.next

    def nextNode(self, l1, l2):
        l1 = l1.next
        l2 = l2.next
        return l1, l2

    def oneBitSum(self, s: int):
        a = s % 10
        b = s // 10
        return [a, b]


nodeL11 = ListNode(9)
nodeL12 = ListNode(8)
nodeL11.next = nodeL12
nodeL21 = ListNode(1)

S = Solution().addTwoNumbers(nodeL11, nodeL21)
while S is not None:
    print(S.val)
    S = S.next
